src/v/cluster/topic_table_stable_iterator.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "topic_table.h"
+
+#include <boost/iterator/iterator_adaptor.hpp>
+
+namespace cluster {
+
+class topics_state_changed_exception final : public std::exception {
+    static constexpr std::string_view ex
+      = "topics map updated, expected revision: {}, current revision: {}";
+
+public:
+    explicit topics_state_changed_exception(
+      model::revision_id expected, model::revision_id current) noexcept
+      : _msg(fmt::format(ex, expected, current)) {}
+
+    const char* what() const noexcept final { return _msg.c_str(); }
+
+private:
+    ss::sstring _msg;
+};
+
+using underlying_const_it = topic_table::underlying_t::const_iterator;
+
+class topics_iterator
```

Comment:
I wonder if we should unify this with `topic_table_partition_generator`

Reply:
I think so but it seems like a invasive change given it is used in multiple places.. probably can be done as a followup so its easy to see if anything breaks.